src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs

```diff
@@ -158,364 +158,430 @@ public Task<RestoreResult> ExecuteAsync()
 
-                    if (!string.IsNullOrEmpty(errorCodes))
-                    {
-                        telemetry.TelemetryEvent[ErrorCodes] = errorCodes;
+                        telemetry.TelemetryEvent[RestoreSuccess] = _success;
                     }
 
-                    if (!string.IsNullOrEmpty(warningCodes))
+                    restoreTime.Stop();
+
+                    // Create result
+                    return new RestoreResult(
+                        _success,
+                        graphs,
+                        checkResults,
+                        msbuildOutputFiles,
+                        assetsFile,
+                        _request.ExistingLockFile,
+                        assetsFilePath,
+                        cacheFile,
+                        cacheFilePath,
+                        packagesLockFilePath,
+                        packagesLockFile,
+                        dependencyGraphSpecFilePath: NoOpRestoreUtilities.GetPersistedDGSpecFilePath(_request),
+                        dependencyGraphSpec: _request.DependencyGraphSpec,
+                        _request.ProjectStyle,
+                        restoreTime.Elapsed);
+                }
+            }
+            catch (Exception ex)
+            {
+                static ILogMessage UnwrapToLogMessage(Exception e)
+                {
+                    var currentException = ExceptionUtilities.Unwrap(e);
+
+                    while ((currentException is FatalProtocolException || currentException is not ILogMessageException) && currentException != null)
                     {
-                        telemetry.TelemetryEvent[WarningCodes] = warningCodes;
+                        currentException = currentException.InnerException;
                     }
 
+                    var logMessageException = currentException as ILogMessageException;
+                    return logMessageException?.AsLogMessage();
+                }
+
+                var contextForProject = CreateRemoteWalkContext(_request, _logger);
 
-                    telemetry.TelemetryEvent[NewPackagesInstalledCount] = graphs.Where(g => !g.InConflict).SelectMany(g => g.Install).Distinct().Count();
+                var unwrappedLogMessage = UnwrapToLogMessage(ex);
 
-                    telemetry.TelemetryEvent[RestoreSuccess] = _success;
+                LockFile assetsFile = BuildAssetsFile(
+                        _request.ExistingLockFile,
+                        _request.Project,
+                        new List<RestoreTargetGraph>(),
+                        new List<NuGetv3LocalRepository>(),
+                        contextForProject);
+
+
+                if (unwrappedLogMessage != null)
+                {
+                    assetsFile.LogMessages.Add(new AssetsLogMessage(LogLevel.Error, unwrappedLogMessage.Code, unwrappedLogMessage.Message, null));
```

Comment:
Your CLI restore example shows that MSBuild is attributing the exception to NuGet.targets, and not the csproj file. There's a property on `AssetsLogMessage` so NuGet tells MSBuild which project file the error is related to, which should be filled in.

Reply:
I just updated the CLI example in the description section. The screenshot I had was an old one - before I started adding the error to an assets file.

Reply:
What about building the assets file *after* you log the message to the logger. 

Currently this adds another place where we can create a discrepancy. 

The pattern currently is:
- You log warnings to the "collector" logger. 
- When you create the assets file, the "collector" logger creates assets file warnings. 
- It'd also be ideal if the failed restore result is created into fewer places. Same reason as above, fewer chances of a discrepancy.

Reply:
We need the assets file, as it’s responsible for capturing all log messages. is the suggestion to do `_logger.Log(error)` instead? or to do `_logger.log(error)` and then follow it up with `asset.LogMessage(error)`

Reply:
No, don't add messages to the assets file. 
There's a collector logger that ensures that all warnings/errors will be part of the built assets file. 

Compare that to how we log errors like NU1008, or warnings like NU1603. You call the logger.

Reply:
Where does the collector logger write the assets file?  This try-catch block is shallow in the call stack, and I wouldn't be surprised if an exception that bubbles all the way up here will not be written to the assets file by the collector logger.

Reply:
I wasn't able to find out if the collector logger writes to the asset file. From my understanding, we do the writing into the assets file in RestoreCommand.ExecuteAsync https://github.com/NuGet/NuGet.Client/blob/46b470cbc5033821d470add040adffe3a1ddf76f/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs#L457

Reply:
You do all the logic restore ever would and you just write to the logger and that's it. 
The try catch should end before we do the transformation at line 457. 

Similarly, don't create a new RestoreResult and keep it in place. That reduces the changes we end up with restore results that don't end up getting everything and either the output is broken or the result is misinterpreted.